File: backend/Logistica/Dominio/Geo.cs
namespace Logistica.Dominio;

/// <summary>
/// Espejo en C# de frontend/lib/dominio/geo.ts. El dispositivo calcula su propio desvío para
/// mostrarlo en el momento de la captura, pero el servidor descarta ese valor y recalcula acá
/// contra la coordenada de la parada — mismo criterio que el precio (construccion_v1.md §3 regla
/// 1): si el número tiene consecuencias (RF-29, marcado sobre umbral), no viene del navegador.
/// Sin API de matriz de distancias, igual que el ruteo: con distancias cortas de reparto urbano,
/// haversine es más que suficiente.
/// </summary>
public static class Geo
{
    private const double RadioTierraM = 6371000;

    public static int DistanciaMetros(decimal latA, decimal lngA, decimal latB, decimal lngB)
    {
        static double Rad(double deg) => deg * Math.PI / 180;

        var dLat = Rad((double)(latB - latA));
        var dLng = Rad((double)(lngB - lngA));
        var senoLat = Math.Sin(dLat / 2);
        var senoLng = Math.Sin(dLng / 2);
        var h = senoLat * senoLat + Math.Cos(Rad((double)latA)) * Math.Cos(Rad((double)latB)) * senoLng * senoLng;
        return (int)Math.Round(2 * RadioTierraM * Math.Asin(Math.Sqrt(h)));
    }
}

File: frontend/lib/dominio/geo.ts
export interface Punto {
  lat: number;
  lng: number;
}

const RADIO_TIERRA_M = 6371000;

/**
 * Distancia en metros entre dos puntos (fórmula de haversine). Sin API de matriz de distancias
 * (construccion_v1.md §1): con 8 a 12 paradas, nearest-neighbor + 2-opt sobre distancia en línea
 * recta queda a pocos puntos del óptimo.
 */
export function haversine(a: Punto, b: Punto): number {
  const rad = (deg: number) => (deg * Math.PI) / 180;
  const dLat = rad(b.lat - a.lat);
  const dLng = rad(b.lng - a.lng);
  const senoLat = Math.sin(dLat / 2);
  const senoLng = Math.sin(dLng / 2);
  const h = senoLat * senoLat + Math.cos(rad(a.lat)) * Math.cos(rad(b.lat)) * senoLng * senoLng;
  return 2 * RADIO_TIERRA_M * Math.asin(Math.sqrt(h));
}
